packages/block-editor/src/components/block-list/block.js

```diff
@@ -111,11 +111,11 @@ function BlockListBlock( {
 
 	const blockType = getBlockType( name );
 	const lightBlockWrapper =
-		blockType.apiVersion > 1 ||
+		blockType?.apiVersion > 1 ||
 		hasBlockSupport( blockType, 'lightBlockWrapper', false );
```

Comment:
Yes, it's for backward compatibility.

> Could we set apiVersion to 2 in a hook during registration to simplify all those checks?

What do you mean?

Reply:
`registerBlockType` could detect if there is `lightBlockWrapper` set in support and when it's `true` then inject `apiVersion` with the version `2`. Would it be compatible?

Reply:
Hmm, have we considered updating the selector for `getBlockType` to return some default not found type, or do we depend on `undefined`?

Normally we should return the `core/missing` block type at this point in block list, but it is possible to unregister the block type. Presumably someone can also break it by updating `setUnregisteredTypeHandlerName` to a nonsense value.

https://user-images.githubusercontent.com/1270189/131414024-8a86d3ad-1ebe-4efd-b6c8-d16e31e379f8.mp4

Reply:
> Hmm, have we considered updating the selector for getBlockType to return some default not found type, or do we depend on undefined?

That would be a breaking change. In the current setup `setUnregisteredTypeHandlerName` is optional for the blocks registry. In practice, it might unconsciously become mandatory. The challenge is that `@wordpress/blocks` can't provide some fallback blocks like the default one, the grouping one, freeform content, and the unregistered type. Maybe that should be the responsibility of the `@wordpress/block-editor` package to define some of those handlers so unregistering a given handler would rollback to the built-in one, or it wouldn't take an effect. We could alternatively refactor code to take into account that those handlers might get unset or the corresponding block type can be removed. @youknowriad and @mtias should know better what was the initial design here and what would be the best way to make it bugs free.

Reply:
That makes sense to avoid a breaking change. If we want to make this a little nicer for callers, one alternative would be to return a default unknown type in the case that the unregistered type handler name is not configured (which supports no properties/has no attributes) in a new selector, say `getBlockTypeOrUnknownBlock`. In its implementation might simply be something like: ```return getBlockType( type ) ?? UNKNOWN_BLOCK_TYPE```. It wouldn't actually be registered, since I think it's fair to add that level of customization.

I think the current design is fine too provided that all callers know that `undefined` may be returned from `getBlockType`. This is easy enough to search for in core usage. We're missing some other usages where we blindly pass `blockType` to another selector, and that selector does not check for undefined. There's also a few spots in code where we also assume that 'core/missing' is registered.

We can of course guard these in follow up PRs:

https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-directory/src/plugins/get-install-missing/install-button.js#L27-L39
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-list/block-html.js#L32-L40
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-list/block.native.js#L321-L323
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-parent-selector/index.js#L41-L52
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-settings-menu/block-mode-toggle.js#L28
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-switcher/block-styles-menu.js#L36
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-switcher/index.js#L79
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-title/index.js#L69
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-tools/block-contextual-toolbar.js#L41-L47
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/block-tools/block-selection-button.js#L79-L85
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/copy-handler/index.js#L34
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/inserter/preview-panel.js#L25
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/inserter/hooks/use-clipboard-block.native.js#L29
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/list-view/block-slot.js#L89
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/components/list-view/list-item.js#L44
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/hooks/align.js#L134-L135
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/block-editor/src/store/actions.js#L764
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/blocks/src/api/factory.js#L290-L291
https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/src/api/serializer.js#L346-L347
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/blocks/src/api/utils.js#L53-L56
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/blocks/src/api/raw-handling/shortcode-converter.js#L95-L98
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/blocks/src/store/selectors.js#L128-L129
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/edit-post/src/hooks/validate-multiple-use/index.js#L111
https://github.com/WordPress/gutenberg/blob/35e16bbad16afb7209fb0e19b87d3a6f4108dd22/packages/edit-site/src/components/header/document-actions/index.js#L27

Reply:
> does `lightBlockWrapper` handling exist only for backward compatibility? I don't see it used with core blocks anymore. I see it referenced in multiple places in the block editor package. Could we set `apiVersion` to `2` in a hook during registration to simplify all those checks?

I opened PR https://github.com/WordPress/gutenberg/pull/34459 that removes usage of `lightBlockWrapper`.